### ISSUE TITLE
[Suggestion] Filter-Upgrade for Relays

### ISSUE THREAD

**Issue by GHXX** [MightyPirates/OpenComputers#2733]
# Why is this suggestion useful?
From what i have seen it is not possible to specify the number of times a packet may pass through relays.
I am assuming this is because people might set it to a number that's extremely high which would cause a massive server load if loops happen. Maybe we could have a card that re-sets the lifetime counter 
(back to 5). This however wouldn't solve the loop issue. 

So what if instead its a card the both resets the counter AND is able to detect looping. If it loops through it 5 times or more its being blocked and not relayed.

# What is my suggestion?
- Add a new upgrade card for relays
- This card will reset the relay-counter so it can be relayed another 5 times. If it goes through another relay with this upgrade the counter will reset again
- To avoid loops there should be a filter functionality built into the card which is able to detect looping packets and blocks them if the very same packet goes through it 5 or more times.

So what do you think about it?

# Alternative solution

An alternative would be to add some sort of custom switch, maybe a network interface or something like that. Basically a micro controller that can interact with network cables. This Block however would have to be able to choose where the packet came from (effectively spoofing the sender address).

**Comment by skyem123** [comment]
Ideally you would wire the system in a way that doesn't loop, although that can get tricky... especially with wireless networks! (maybe that is something else to wonder about...)

The way I would rather it be done is by allowing people to make their own switches by being able to sniff and spoof packets, but other people here don't like my idea. :p

Maybe the "card" could instead be a fully programming thing, similar to a microcontroller? That would have some interesting implications...

Sorry for kinda hijacking this...
oh also why 5 times, specifically?

**Comment by GHXX** [comment]
as the wiki (iirc) and fnuecke say:

> Sounds like you have a loop in your network, in which case that's expected and intended behavior. Network packets have a TTL of 5 hops, so this sounds like the switch is forwarding it those 5 times.

But yeah, custom relays could be very cool actually. Ill update the issue.

**Comment by skyem123** [comment]
oh I see now, a card that does more detailed inspection of packets, while also resetting the TTL, I think it's not even needed to have the the limit as high as 5?
The main problem I see is that what if people do send the same message over and over and want their data back? That would require every packet to have a hidden unique ID which is... eeeh.

**Comment by GHXX** [comment]
or there could be a hidden, salt (a value that counts up converted to letters for example) added to the packet. And a TTL of 5 wont hurt. So you dont need a special switch every hop.

**Comment by GHXX** [comment]
Or there could just be a "No TTL decrease card", that tells the relay not to decrease the TTL. If you build a loop then your relay backlogs because you messed up 😄

**Comment by Vexatos** [comment]
This sounds like something that should be handled by a Microcontroller. Of course, these would only be able to receive (and then forward) messages sent to them either directly or via `broadcast`, but that, in turn, would give someone a reason to write a smart and potentially secure networking protocol. I think the set amount of hops is a good way to limit the usefulness of `modem.send` compared to `modem.broadcast`.

**Comment by GHXX** [comment]
People could come up with a proper minecraft-http like protocol and make a library for communication over microcontrollers that could also resolve hostnames for example (ingame). Would be interesting for sure! I wonder what @payonel thinks about this.

To summarise all of this:
The basic problem is that the TTL of network packets is 5 and that it cannot be increased easily. There are a few possible solutions (first: best, last: worst (in my opinion)):

1.  Specialised network-microcontrollers that can be programmed to do what you want them to do. They are like microcontrollers, but they can only handle networking. Every side could have its own network card maybe? Not too sure... Maybe it could have an upgrade to also pass energy through the cable. No other upgrades though i think. (would allow people to make smart switches)
2.  An upgrade card for relays that tracks all packets and filters out looping ones. When a packet passes through a relay with this card the TTL is reset to 5. (Might require tinkering in lots of areas since it would need every packet to have an ID) (Might cause further issues down the line. more complicated code...)
3.  Just place a computer with multiple network cards and .... Thats just overkill and wouldn't work well for bigger networks i think. (Overkill, very expensive in resources as in iron, gold, etc.)

**Comment by skyem123** [comment]
The main problem with computers with different network cards and microcontrollers as they are now is that they are not transparent to other devices so everything has to have the higher layer protocol implemented for it to be useful... :p

**Comment by GHXX** [comment]
yeah, but you would just put it onto a floppy disk and install it :P

**Comment by xarses** [comment]
out of curiosity, why do you want the packet to loop more than 5 times through the relay? Seems like the issue is there, not with needing to reset the TTL

**Comment by GHXX** [comment]
What if you have huge nerworks? Chances are that a TTL of 5 isnt enough.....

**Comment by GHXX** [comment]
So, it has been a while, @payonel what do you think about this?
(Adding a new microcontroller that can only interact with networks to act as a network switch)

**Comment by xarses** [comment]
I'm still at a loss for a reasonable construction that needs a grater than 5 ttl.

P.S. You can already use a computronix spoofing card to act as a switch

**Comment by skyem123** [comment]
@xarses, I did check a few months ago, and unless that changed, you cannot act as a switch with the spoofing card, as you cannot listen to packets not sent to its.

**Comment by GHXX** [comment]
Tbh, you reach 5 bounces quite quickly. Say you got an elevator, thats controlled by a server. Every computer needs to be separated by a relay. so thats two relays already (if the server also gets one) When you now connect that to a different network you may want another relay, to keep it clean. You also may want to prevent power from flowing through the network. Then you also use a relay. 
Imagine a bigger network now, where say a hundred pcs connect and interact with eachother, controlling stuff...

TL;DR You reach it quite quickly. Yeah sure, you could place a computer and then broadcast everything. And then you can use the pull event to listen for it. What if two packets arrive at the same time? One gets dropped. What if you want to send it to a specific address? (as @skyem123 mentioned) 
And doing it like that is pretty expensive.
What if you dont want someone to read the broadcasted packets? Is that where PKI kicks in? I think that would be overkill, even though its kinda necessary at that point.
TL;DR # 2 Workarounds are possible, yet not practical.
